<p>I have been working on single-cell data for almost a year now and I will be using this blog to detail what I have learned along with some tips and tricks.</p>

<p>My research focuses on development of mesenchymal tissues and how genes that regulate these processes could be used to better understand soft tissue sarcomas and osteosarcomas.</p>

<p>Below are some publications, tutorials, and workshops that I found helpful.</p>

<h2 id="publications">Publications</h2>

<ul>
  <li>
    <p><a href="https://www.cell.com/cell/pdf/S0092-8674(19)30559-8.pdf">Comprehensive Integration of Single-Cell Data</a> - This article comes from the <a href="https://satijalab.org/">Satija lab</a>, who developed <a href="https://satijalab.org/seurat/">Seurat</a>, one of the most popular R package for single-cell RNA-seq data. This paper highlights Seurat V3 which added methods for single-cell integration, normalization using sctransform, as well as a restructured Seurat object for multi-modal data (i.e., RNA, ATAC, protein, etc.). I highly recommend reading and re-reading this article as I always find myself learning something new.</p>
  </li>
  <li>
    <p><a href="https://www.embopress.org/doi/10.15252/msb.20188746">Current best practices in single‐cell RNA‐seq analysis: a tutorial</a> - This is a must read for anyone entering the single-cell space. It details all of the current best practices while providing rationale and many alternate methods. It answered a majority of my questions when the tutorials I previously read did not give a reason as to why such methods were being performed on the data.</p>
  </li>
</ul>

<h2 id="tutorials">Tutorials</h2>

<ul>
  <li>
    <p><a href="https://satijalab.org/seurat/vignettes.html" target="_blank">Seurat</a></p>
  </li>
  <li>
    <p><a href="https://cole-trapnell-lab.github.io/monocle3/docs/introduction/" target="_blank">Monocle 3</a></p>
  </li>
  <li>
    <p><a href="https://scanpy.readthedocs.io/en/stable/tutorials.html" target="_blank">Scanpy</a></p>
  </li>
</ul>

<h2 id="workshops">Workshops</h2>

<ul>
  <li>
    <p><a href="https://bioc2020.bioconductor.org/workshops" target="_blank">BioC 2020: Where Software and Biology Connect</a> - Added 2020/07/19. Over 20 workshops related to single-cell analysis.</p>
  </li>
  <li>
    <p><a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html" target="_blank">Orchestrating Single-Cell Analysis with Bioconductor</a> - This is a very in-depth and comprehensive workshop for single-cell RNA-seq analysis in R using packages from Bioconductor. I have not yet gone through it because I saw it today and it was posted online 2 days ago.</p>
  </li>
  <li>
    <p><a href="https://broadinstitute.github.io/2019_scWorkshop/" target="_blank">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a> - This workshop may be a bit outdated now but it runs through the entire workflow of single-cell RNA-seq processing and analysis. I found it incredibly useful as not only does it provide many alternate methods for processing, this workshop also includes the code to run the different packages too.</p>
  </li>
  <li>
    <p><a href="https://scrnaseq-course.cog.sanger.ac.uk/website/index.html" target="_blank">Analysis of single cell RNA-seq data</a> - The above workshop and this one may be outdate as well since both were published in 2019. Nonetheless, they still contain useful information including comparisons between different analytical methods, albeit perhaps outdated as well.</p>
  </li>
  <li>
    <p><a href="https://satijalab.org/scgd/" target="_blank">Single Cell Genomics Day: A Practical Workshop</a> - Definitely keep an eye on this as it is updated yearly. This is in-person workshop but the presentations are recorded (some of them) and the slides are posted. It is a good place to check out if you are interested in seeing how the latest techniques are being used.</p>
  </li>
</ul>

<p>That’s it for now. I may update this if I see new information.</p>

